src/EditorFeatures/Core/Shared/Extensions/OperationContextExtensions.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.ErrorReporting;
+using Microsoft.CodeAnalysis.Host;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Shared.Extensions
+{
+    internal static class OperationContextExtensions
+    {
+        /// <summary>
+        /// Wait until given document is fully loaded and then add scope
+        /// </summary>
+        public static IUIThreadOperationScope WaitForDocumentAndAddScope(
+            this IUIThreadOperationContext context, Document document, bool allowCancellation, string description)
+        {
+            return WaitAndAddScope(context, document.Project.Solution, document.Name, allowCancellation, description);
+        }
+
+        public static IUIThreadOperationScope WaitForProjectAndAddScope(
+            this IUIThreadOperationContext context, Project project, bool allowCancellation, string description)
+        {
+            return WaitAndAddScope(context, project.Solution, project.Name, allowCancellation, description);
+        }
+
+        private static IUIThreadOperationScope WaitAndAddScope(
+            this IUIThreadOperationContext context, Solution solution, string name, bool allowCancellation, string description)
+        {
+            IUIThreadOperationScope scope = null;
+
+            try
+            {
+                // partial mode is always cancellable
+                var title = string.Format(EditorFeaturesResources.Operation_is_not_ready_for_0_yet_see_task_center_for_more_detail, name);
+                scope = context.AddScope(allowCancellation: true, title);
```

Comment:
this seems strange.  What if the caller said "i don't allow cancellation" (which was necessary for correctness in their domain).  It now seems like we woudl allow cancellation on this sub scope.  The user could then cancel, and break the invariant of the calling code which said that cancellation was not allowed.

Reply:
waiting on partial mode is always allowed.

Reply:
> cancellation on waiting on partial mode is always allowed.

This may break the expectation of callers.  Again, say we have a caller that passes CancellationToken.None, signifiying that they cannot be canceled, and may break invariants if cancellation happens, this will now violate their design.

This is not hypothetical.  We have codepaths taht pass along CancellationToken.None because htey explicitly do not support cancellation happenin from that point on (i.e. they've mutated some state, and need further changes to happen).  If you now throw a cancellation exception, this can break them.

this is also an issue since you have a new cancellation token here.  If you throw off of this, you could cause faulted tasks if hte task infrastructure was passed another cancellation token.  This is an unsafe thing you are doing here.

Reply:
it will discuss that when that happen since I dont believe that can ever happen for this code path.